chainstate/tx-verifier/src/transaction_verifier/hierarchy.rs

```diff
@@ -75,6 +77,16 @@ where
         self.storage.get_gen_block_index(block_id)
     }
 
+    fn get_undo_data(
+        &self,
+        tx_source: TransactionSource,
+    ) -> Result<Option<CachedUtxosBlockUndo>, <Self as TransactionVerifierStorageRef>::Error> {
+        match self.utxo_block_undo.data().get(&tx_source) {
+            Some(op) => Ok(op.get().cloned()),
```

Comment:
Why not return a reference instead of cloning?

Reply:
We are mimicking storage interface here and return everything by value